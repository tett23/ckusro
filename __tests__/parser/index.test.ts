import {
  CkusroFile,
  FileTypeDirectory,
  FileTypeDoesNotExist,
  LoaderContext,
} from '../../src/loader';
import {
  buildAst,
  buildDoesNotExistFile,
  determineDependency,
} from '../../src/parser';
import { buildFile, buildLoaderContext } from '../__fixtures__';

describe(buildAst, () => {
  it('builds AST', () => {
    const actual = buildAst('# hoge');

    expect(actual).toMatchSnapshot();
  });

  it('paeses wikiLink', () => {
    const actual = buildAst('foo[[foo]]bar');

    expect(actual).toMatchSnapshot();
  });
});

describe(buildDoesNotExistFile, () => {
  const expected: CkusroFile = {
    id: 'test:/does_not_exist',
    namespace: 'test',
    name: 'does_not_exist',
    path: '/does_not_exist',
    fileType: FileTypeDoesNotExist,
    isLoaded: false,
    content: null,
    weakDependencies: [],
    strongDependencies: [],
    variables: [],
  };

  it('returns FileTypeDoesNotExist object', () => {
    const actual = buildDoesNotExistFile('test', '/does_not_exist');

    expect(actual).toEqual(expected);
  });

  it('transforms path into absolute path', () => {
    const actual = buildDoesNotExistFile('test', 'does_not_exist');

    expect(actual).toEqual(expected);
  });
});

describe(determineDependency, () => {
  it('returns CkusroIds', () => {
    const context = buildLoaderContext({ name: 'test', path: '/test' });
    const rootNode = buildAst('[[test:foo]]');
    const files: CkusroFile[] = [
      buildFile({
        namespace: 'test',
        name: 'foo',
        path: '/foo',
        fileType: FileTypeDirectory,
        isLoaded: false,
        content: null,
        weakDependencies: [],
        strongDependencies: [],
      }),
    ];
    const actual = determineDependency(context, rootNode, files);

    expect(actual).toEqual([files[0]]);
  });

  it('returns empty array when file does not exist', () => {
    const context: LoaderContext = {
      name: 'test',
      path: '/test',
    };
    const rootNode = buildAst('[[test:does_not_exist]]');
    const files: CkusroFile[] = [];
    const actual = determineDependency(context, rootNode, files);

    expect(actual).toMatchObject([
      {
        id: 'test:/does_not_exist',
        namespace: 'test',
        name: 'does_not_exist',
        path: '/does_not_exist',
        fileType: FileTypeDoesNotExist,
        isLoaded: false,
        content: null,
        weakDependencies: [],
        strongDependencies: [],
        variables: [],
      },
    ]);
  });
});
