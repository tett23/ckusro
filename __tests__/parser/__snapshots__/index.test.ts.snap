// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildAst builds AST 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 7,
              "line": 1,
              "offset": 6,
            },
            "indent": Array [],
            "start": Object {
              "column": 3,
              "line": 1,
              "offset": 2,
            },
          },
          "type": "text",
          "value": "hoge",
        },
      ],
      "depth": 1,
      "position": Position {
        "end": Object {
          "column": 7,
          "line": 1,
          "offset": 6,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "heading",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`buildAst paeses wikiLink 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "foo",
        },
        Object {
          "data": Object {
            "component": "WikiLink",
            "props": Object {
              "internalLink": Object {
                "target": "foo",
              },
              "text": "foo",
            },
          },
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "indent": Array [],
            "start": Object {
              "column": 4,
              "line": 1,
              "offset": 3,
            },
          },
          "type": "jsx",
          "value": "<WikiLink linkTarget=\\"foo\\" className=\\"\\">foo</WikiLink>",
        },
        Object {
          "position": Position {
            "end": Object {
              "column": 14,
              "line": 1,
              "offset": 13,
            },
            "indent": Array [],
            "start": Object {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
          },
          "type": "text",
          "value": "bar",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 14,
          "line": 1,
          "offset": 13,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 14,
      "line": 1,
      "offset": 13,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
